File: 002_AddTwoNumbers/002-add-two-numbers.cpp
/*
Problem : https://leetcode.com/problems/add-two-numbers/description/
Add two (reverse order) numbers and return it as a linked list.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8

2 4 3
5 6 4
-----
7 0 8(carry + 1)
*/

struct ListNode {
     int val;
     ListNode *next;
     ListNode(int x) : val(x), next(nullptr) {}
};

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
     
        ListNode* res = new ListNode(0);
        ListNode* prev = res;
        int carriage = 0;
        
        while(true)
        {
            carriage = linkAddTwoNumbers(prev, l1, l2);
            l1 = l1 != nullptr ? l1->next : nullptr;
            l2 = l2 != nullptr ? l2->next : nullptr;
            if (l1 != nullptr || l2 != nullptr || carriage != 0)
            {
                prev->next = new ListNode(carriage);
                prev = prev->next;
            }
            else
            {
                break;
            }
        }
        
        return res;
    }

    int linkAddTwoNumbers(ListNode* res, ListNode* l1, ListNode* l2)
    {
        int sum = res->val;
        if (l1 != nullptr)
        {
            sum += l1->val;
        }
        if (l2 != nullptr)
        {
            sum += l2->val;
        }
        
        res->val = sum % 10;
        
        return sum / 10;
    }
};

File: python/002-add-two-numbers.py
"""
Problem : Add Two Numbers (https://leetcode.com/problems/add-two-numbers/description/)
"""

# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    """Solution Class"""

    def addTwoNumbers(self, l1, l2):
        """
        :type l1: ListNode
        :type l2: ListNode
        :rtype: ListNode
        """
        carry = 0
        result = ListNode(0)
        next_result = result
        while True:
            next_result.val, carry = Solution.add_nodes(l1, l2, carry)
            l1 = l1.next if l1 != None else None
            l2 = l2.next if l2 != None else None

            if l1 == None and l2 == None and carry == 0:
                break

            prev_result = next_result
            next_result = ListNode(0)
            prev_result.next = next_result

        return result
    @staticmethod
    def add_nodes(l1, l2, carry):
        sum_numbers = carry
        if l1 != None:
            sum_numbers += l1.val
        if l2 != None:
            sum_numbers += l2.val
        return [sum_numbers % 10, sum_numbers // 10]